src/librustc_mir/build/scope.rs

```diff
@@ -94,17 +94,31 @@ use rustc::ty::subst::{Kind, Subst};
 use rustc::ty::{Ty, TyCtxt};
 use rustc::mir::*;
 use rustc::mir::transform::MirSource;
-use syntax_pos::Span;
+use syntax_pos::{self, Span};
 use rustc_data_structures::indexed_vec::Idx;
 use rustc_data_structures::fx::FxHashMap;
 
+#[derive(Debug)]
+pub struct DivergePathState {
+    pub(crate) block: BasicBlock,
```

Comment:
It would be really great to put a comment on this struct explaining the overall flow. e.g., when is `terminator_emitted` set to true, etc? Clearly, this is encoding some sort of state machine, but it's a somewhat opaque one.